util/timer.h

```diff
@@ -212,14 +242,13 @@ class Timer {
   std::unique_ptr<port::Thread> thread_;
   bool running_;
 
-
-  std::priority_queue<FunctionInfo*,
-                      std::vector<FunctionInfo*>,
-                      RunTimeOrder> heap_;
+  std::priority_queue<std::shared_ptr<FunctionInfo>,
+                      std::vector<std::shared_ptr<FunctionInfo>>, RunTimeOrder>
+      heap_;
 
   // In addition to providing a mapping from a function name to a function,
   // it is also responsible for memory management.
-  std::unordered_map<std::string, std::unique_ptr<FunctionInfo>> map_;
+  std::unordered_map<std::string, std::shared_ptr<FunctionInfo>> map_;
```

Comment:
Why do `map_` and `heap_` share ownership of the `FunctionInfo` objects?

Reply:
It's to handle re-add a function with the same name, like:

Reply:
Implementation details aside, I don't think `map_` and `heap_` are equal, thus they should not share ownership equally. `map_` manages the resources, e.g. memory allocated for `FunctionInfo` objects, while `heap_` should just contain light references to these functions.
With #7228, it may be easier to make sure the situation described does not happen: when a function is removed from `map_`, it should also be removed from `heap_`. When cancelling a function, the function should be removed from the queue already.

Reply:
As we discussed offline, if there's an existing function with the same name, we will just override it.